opencodelists/tests/functional/test_build_codelist.py

```diff
@@ -0,0 +1,583 @@
+        assert False, "Coding system {coding_system} not yet supported by the tests"
+
+    page.get_by_role("link", name="My codelists").click()
+    page.get_by_role("link", name="Create a codelist!").click()
+
+    page.get_by_role("textbox", name="Codelist name *").fill(name)
+    page.get_by_label("Coding system *").select_option(coding_system)
+    page.get_by_role("button", name="Create").click()
+
+
+def search_in_codelist_builder(page, search_action):
+    """Takes a Playwright page that's navigated to the codelist builder,
+    and str search.
+    Performs the provided search on the codelist builder page."""
+
+    # This selection depends on the limited hierarchy that we have always
+    # using decimal codes, and non-decimal names.
+    #
+    # WARNING: this holds for our SNOMED-CT and BNF fixtures,
+    # but definitely does not hold for the ICD10 fixtures,
+    # though we are not testing any ICD10 data as yet.
+    if search_action.search[0].isdecimal():
+        page.get_by_role("radio", name="Code").click()
+
+    page.get_by_role("searchbox").click()
+    page.get_by_role("searchbox").fill(search_action.search)
+    page.get_by_role("button", name="Search", exact=True).click()
+
+
+def handle_concept_selections(page, search_action):
+    """Takes a Playwright page that's navigated to the codelist builder,
+    and a SearchAction.
+    Perform the appropriate ConceptSelection actions belonging to that
+    SearchAction on the Playwright page."""
+
+    for concept_selection in search_action.concept_selections:
+        concept_locator = page.locator(f"[data-code='{concept_selection.code}']")
+
+        if concept_selection.to_expand:
+            concept_locator.get_by_role("button", name="⊞").click()
+
+        match concept_selection.state:
+            case ConceptState.INCLUDED:
+                button_icon = "+"
+            case ConceptState.EXCLUDED:
+                button_icon = "−"
+            case _:
+                # Guard against adding another ConceptState value.
+                assert False
+
+        concept_locator.get_by_role("button", name=button_icon).click()
+
+
+def save_codelist_for_review(page):
+    """Takes a Playwright page on a codelist builder page,
+    with a completed codelist, and saves the codelist for review."""
+
+    expect(page.get_by_role("button", name="Save for review")).to_be_enabled()
+    page.get_by_role("button", name="Save for review").click()
+
+
+def validate_codelist_exists_on_site(page, codelist_name, heading_text):
+    """Takes a Playwright page, a codelist name, and heading text to match
+    and validates that the codelist name exists under that heading.
+
+    It does not currently go for an exact match because organisation users get
+    "owned by organisation" appended to the codelist name.
+
+    We could improve this matching in future."""
+
+    page.get_by_role("link", name="My codelists").click()
+    # TODO: can we remove the duplication here?
```

Comment:
Do you mean the duplication of `page.get_by_role("link", name=codelist_name)`? In which case you can just do `codelist_link = page.get_by_role("link", name=codelist_name)` - but maybe that's not what you meant.

Reply:
Specifically, I think my question was: is it OK/idiomatic to assign such a thing you'll then check with `expect`, before you do that check?

Reply:
Assigning locators is fine. There's a section in [the docs](https://playwright.dev/python/docs/locators#locating-elements) about it with examples. Also the locator reevaluates each time the assigned variable is used.